Field Filter Value Normalization in Embedding Endpoints

adam-james-v opened the pull request:
Fixes #27643

The issue actually mentions 2 different bugs:
 - the card embedding preview was failing. This is due to `"true"` not being `true`, so the normalization change fixes this.
 - the dashboard embedding preview was failing immediately AND the filter wasn't presenting a list, just an input box. The first immediate failing was due to no `:embedding-parameters` existing in the Appdb, upon 'Publish' that issue went away, so I fixed the param-values implementation to address this case. The second issue was the same as with card embedding, so that's fixed as well.

Since we use query-params to pass paramter values, we have to normalize the parameter map values.

This PR is a WIP, as I'm trying to find a way to do this hollistically/properly.

First commit here is mostly as a marker for where to start working on this properly, as I think the json parsing won't catch all cases, nor should the normalization happen without some attempt to use a schema to properly validate, maybe even coerce things.

I also want to build a small set of tests to catch these differences across cards, dashboards, card embeds, and dashboard embeds.

Finally, this may close other issues, but I still have to confirm that.

adam-james-v committed:
Field Filter Value Normalization in Embedding Endpoints

Fixes 27643

Since we use query-params to pass paramter values, we have to normalize the parameter map values.

This PR is a WIP, as I'm trying to find a way to do this hollistically/properly.

First commit here is mostly as a marker for where to start working on this properly, as I think the json parsing won't
catch all cases, nor should the normalization happen without some attempt to use a schema to properly validate, maybe
even coerce things.

adam-james-v committed:
Merge branch 'master' into bugfix-27643-field-filters-embedding-normalization

adam-james-v committed:
read strings if boolean or numbers, replace blank str with nil

Also update the common dashboard param values fn to use embedding params from the URL blob if the app db doesn't have
embedding params. This happens upon first preview where nothing will have been written to the Appdb yet.

adam-james-v committed:
Add query-param normalization test, add related embedding test

adam-james-v commented on src/metabase/api/embed/common.clj:
```diff
@@ -125,9 +125,22 @@
-  (-> query-params
-      (update-keys keyword)
-      (update-vals (fn [v] (if (= v "") nil v)))))
+  (letfn [(maybe-read [v]
+            (if (string? v)
+              (cond
+                (#{"true" "false"} v) (parse-boolean v)
+                (str/blank? v)        nil
+                :else                 (or (parse-long v)
+                                          (parse-double v)
+                                          v))
+              v))]
+    (-> query-params
+        (update-keys keyword)
+        (update-vals (fn [v]
+                       (if (and (not (string? v))
+                                (seq v))
+                         (mapv maybe-read v)
+                         (maybe-read v)))))))
```
I only really care about turning numbers and booleans into their expected types, anything else can be kept as strings.

query-params come from the URL, like this: `/api/embedding_endpoints/some_token?my_param=true&my_param2=asdf&my_param2=wasd`

So I want that to end up as:
```
{:my_param true
 :my_param2 ["asdf" "wasd"]}
 ```
 
 And this normalize takes care of that. There's a test added to show this behaviour as well.

adam-james-v committed:
schema didn't work in this case after all.

Worth a shot, but don't want this to block the PR, so reverting to the previous schema

adam-james-v commented on src/metabase/api/dashboard.clj:
```diff
@@ -1225,10 +1225,10 @@
+  {dashboard-id ms/PositiveInt
+   dashcard-id  ms/PositiveInt
+   card-id      ms/PositiveInt
+   parameters   [:maybe [:sequential ms/Parameter]]}
```
Well, nevermind for now. Don't want this schema to block the PR, so I'll revert for now.

tsmacdonald commented on src/metabase/api/embed/common.clj:
```diff
@@ -125,9 +125,22 @@
+  (letfn [(maybe-read [v]
+            (if (string? v)
+              (cond
+                (#{"true" "false"} v) (parse-boolean v)
```
I don't have any brilliant suggestions, but it seems crazy that this is the first time we've had to deal with this? I'm surprised there's not a lib function somewhere that replicates this entire `cond`

tsmacdonald replied:
@escherize are we missing a trick here with using `malli.schema.BooleanValue` and friends?

adam-james-v replied:
Yea, I'd also gladly use an existing util fn or something else if there's something obvious!

escherize replied:
I'm not sure we should be converting every single "true" query parameter value to be `true`, shouldn't that be driven by the arg->schema map in defendpoint instead? It would seem like a surprise if I chose true or false for a collection/db/query/etc name, (that got sent through a qp) this could break when e.g. inserting `true` where we rightly expect "true".

If we do it this way we should add some docs that say you're not allowed to represent the strings true or false as query param values

escherize replied:
> are we missing a trick here with using malli.schema.BooleanValue and friends?

Yeah, I think we should just coerce the query param value inside of defendpoint.

escherize replied:
I think the mechanim to convert these doesn't matter. I was under the impression that there is a way to go from the name of the key to the expected type, and I was going to advocate for deriving the proper conversion function that way. @adam-james-v has informed me that there actually isn't a good way to know what type a value in these query params should actually be, so I think this guess is fine, and should fix the P1.

We should probably revisit this and figure out how to actually type the query param data.

tsmacdonald commented on test/metabase/api/preview_embed_test.clj:
```diff
@@ -577,3 +577,62 @@
+  (testing "embedding endpoint parameter values are parsed when sensible. (#27643)"
+    (embed-test/with-embedding-enabled-and-new-secret-key
+      (mt/dataset places-cam-likes
+        (mt/with-temp [:model/Card {card-id :id :as card} {:dataset_query
```
`mt/native-query` would save you a couple tokens of typing, but NBD

adam-james-v replied:
TIL. 
I'll leave this be, but remember that for future tests :)

tsmacdonald reviewed (approved):
Behavior looks reasonable to me (but I Am Not An Embedding Dev)

adam-james-v committed:
Add comment pointing out the downsides/reasons for this solution

It's nto a perfect solution, but it's probably the best we can get at the moment, without adding a new filter
type (boolean) and/or improving the design of hte embedding endpoints and how filters/parameters work
